jax/lax/lax_control_flow.py

```diff
@@ -974,14 +976,194 @@ def _root_jvp(primals, tangents, num_consts, jaxpr, solve, tangent_solve):
   return solution, solution_dot
 
 
-def _root_batch(args, dims, **params):
-  return batching.batch_fun(lu.wrap_init(_root_impl, params), args, dims)
-
-
 root_p = core.Primitive('root')
 root_p.multiple_results = True
 root_p.def_impl(_root_impl)
 root_p.def_abstract_eval(_root_abstract_eval)
 ad.primitive_jvps[root_p] = _root_jvp
 xla.initial_style_translations[root_p] = xla.lower_fun(_root_impl, initial_style=True)
-batching.primitive_batchers[root_p] = _root_batch
+# TODO(shoyer): write batching rule
+
+
+class _LinearSolveTuple(collections.namedtuple(
+    '_LinearSolveTuple', 'matvec, vecmat, solve, transpose_solve')):
+
+  def transpose(self):
+    return type(self)(self.vecmat, self.matvec, self.transpose_solve, self.solve)
+
+
+def _split_linear_solve_args(args, const_lengths):
+  params_list = split_list(args, list(const_lengths))
+  return _LinearSolveTuple(*params_list[:-1]), params_list[-1]
+
+
+def _transpose_function(linear_fun, primals):
+  """Transpose a linear function."""
+  # TODO(shoyer): can we use something more direct than the vjp machinery?
+  # It's particularly awkward that we need the second argument to give
+  # particular values of the primals, which are entirely arbitrary.
+  _, vjp_fun = api.vjp(linear_fun, primals)
+
+  def transposed_fun(x):
+    (y,) = vjp_fun(x)
+    return y
+
+  return transposed_fun
+
+
+def _flatten(args):
+  return [x for arg in args for x in arg]
+
+
+def _check_shapes(func_name, expected_name, actual, expected, tree):
+  actual_shapes = _map(onp.shape, actual)
+  expected_shapes = _map(onp.shape, expected)
+  if actual_shapes != expected_shapes:
+    actual_shape_tree = tree_unflatten(tree, actual_shapes)
+    act_shape_tree = tree_unflatten(tree, actual_shapes)
+    raise ValueError('{}() output shapes must match {}, got {} and {}'
+                     .format(func_name, expected_name,
+                             tree_unflatten(tree, actual_shapes),
+                             tree_unflatten(tree, expected_shapes)))
+
+
+def linear_solve(matvec, b, solve, transpose_solve=None, symmetric=False):
```

Comment:
Does it make sense to call it `linear_solve` if the user has to provide a function to solve the linear system? The purpose of this function is clear to me, but I'm trying to put myself in a new user's shoes.

A new user might not be familiar with the subtleties of automatic differentiation. This is similar to our `lax.root` discussion. It might be good to make sure it is clear that the purpose is to define "appropriate" derivatives. I think that something with `def*` prefix, ala `defjvp`, makes it clear but the docs alone could possibly do the trick. I'm afraid that a new user would read `linear_solve` and just assume to know what it does. An "unusual" name (like something with a `def*` prefix) would avoid users assuming they know what the function does just from the name.
WDYT?

Reply:
I agree that an unusual name would be a good idea here. I'm leaning towards `differentiable_linear_solve` per our discussion over in https://github.com/google/jax/issues/1448. (I'll also definitely add clearer docs explaining why this is useful!)

Reply:
Maybe `custom_linear_solve`?

Reply:
Our discussion is now a little split but I'm on board!

> I'll also definitely add clearer docs explaining why this is useful!

Great! Though, there is no need to over do it here, imo. 2-3 brief sentences should be enough for people to know what to look up if they don't understand. More detailed docs might deserve a standalone implicit diff page similar to async dispatch and co, that explains why backprop through all the things isn't the only solution.